libs/datasets/vaccine_backfills.py

```diff
@@ -35,14 +36,30 @@ def derive_vaccine_pct(ds_in: MultiRegionDataset) -> MultiRegionDataset:
     )
     derived_pct_df = derived_pct_df.rename(index=field_map, level=PdFields.VARIABLE)
 
-    # TODO(tom): Compare derived_pct_df and ts_in_pcts to pick least stale.
-    # Drop derived percentages where ts_in_pcts already has a timeseries with the same index.
-    derived_pct_without_in_pcts = derived_pct_df.loc[~derived_pct_df.index.isin(ts_in_pcts.index)]
-    # Double check that there is no overlap between the percentages that will be added and the
-    # existing data.
-    assert ts_in_all.index.intersection(derived_pct_without_in_pcts.index).empty
+    def append_most_recent_date_index_level(df: pd.DataFrame) -> pd.DataFrame:
+        """Appends most recent date with real (not NA) value as a new index level."""
+        most_recent_date = df.apply(pd.Series.last_valid_index, axis=1)
+        return df.assign(most_recent_date=most_recent_date).set_index(
```

Comment:
Just when I think I understand enough python syntax to get by, I see `most_recent_date=most_recent_date` and have no idea how that works. 😬

Reply:
This is calling a function with the key of a keyword parameter equal to a local var, actually pretty common:

```
$ find -name '*py' | xargs grep -P -r -e '\((\w+)=\1,'
./tests/libs/series_utils_test.py:    smoothed = series_utils.smooth_with_rolling_average(series=series, window=2)
./tests/libs/series_utils_test.py:    smoothed = series_utils.smooth_with_rolling_average(series=series, window=5)
./cli/data.py:                dataset = dataset.get_subset(state=state, fips=fips)
./pyseir/cli.py:    for state in _states_region_list(state=state, default=ALL_STATES):
./pyseir/models/suppression_policies.py:    return interp1d(x=x, y=y, fill_value="extrapolate")
./libs/github_utils.py:    artifact_url, run_number = _get_artifact_zip_url(run_number=run_number, latest=latest)
./libs/datasets/data_version.py:            return func(version=version, **kwargs)
...
```
But I find the usage with `DataFrame.assign` a bit goofy because the parameter name is turned into a column label of the dataframe. I think it'd be better having a real dict, ie `df.assign({'most_recent_date': most_recent_date)...`

Reply:
Ohh....  I see.  Yeah, I knew Python had a version of named arguments but I didn't fully realize that it could essentially just be treated as an arbitrary dict under the covers...  and so with `.assign()`, the lefthand side can be an arbitrary column name as opposed to an actual function argument, so it didn't click that that's the mechanism being leveraged here... Makes sense now.

Yeah. FWIW, it seems like `.assign()` does accept an actual dict as well, so leaning towards the the `.assign({'most_recent_date': most_recent_date})` syntax as a convention in our code might be reasonable. 😁   Anyway, thanks for the pointer.

Reply:
I don't think `.assign({'most_recent_date': most_recent_date})` works. It raises `TypeError: assign() takes 1 positional argument but 2 were given` (the first positional arg is the implicit `self` and the second is the dict). `.assign(**{'most_recent_date': most_recent_date})` works but depends on a different slightly esoteric python calling syntax.

Reply:
Gah, sorry. You're of course right.  I just glanced at the [docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.assign.html) and the first example had `{'temp_c': [17.0, 25.0]}` in it, but looking closer, that was actually a call to `pd.DataFrame()` to set up an example dataframe. Sorry! 😬